solr/src/main/java/org/apache/metamodel/solr/SolrDataContext.java

```diff
@@ -0,0 +1,429 @@
+import org.apache.solr.client.solrj.SolrQuery.ORDER;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.impl.HttpSolrServer;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.client.solrj.response.FacetField;
+import org.apache.solr.common.SolrDocumentList;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+
+/**
+ * DataContext implementation for ElasticSearch analytics engine.
+ * 
+ * ElasticSearch has a data storage structure hierarchy that briefly goes like
+ * this:
+ * <ul>
+ * <li>Index</li>
+ * <li>Document type (short: Type) (within an index)</li>
+ * <li>Documents (of a particular type)</li>
+ * </ul>
+ * 
+ * When instantiating this DataContext, an index name is provided. Within this
+ * index, each document type is represented as a table.
+ * 
+ * This implementation supports either automatic discovery of a schema or manual
+ * specification of a schema, through the {@link SimpleTableDef} class.
+ */
+
+public class SolrDataContext extends QueryPostprocessDataContext implements
+		DataContext {
+
+	private static final Logger logger = LoggerFactory
+			.getLogger(SolrDataContext.class);
+
+	public static final String FIELD_ID = "_id";
+
+	private final SimpleTableDef tableDef;
+	private final String indexName;
+	private final String url;
+
+	private SolrQuery query = new SolrQuery();
+
+	private static enum queryTypes {
+		ROW, FACET
+	};
+
+	private static final int MAX_RETRIES = 0;
+	private static final int SOCK_TIMEOUT = 1000;
+	private static final int CONN_TIMEOUT = 5000;
+	private static final int DEFAULT_LIMIT = 1000;
+
+	public SolrDataContext(String url, String indexName) {
+		this.url = url;
+		this.indexName = indexName;
+		this.tableDef = detectSchema(url, indexName);
+	}
+
+	private SimpleTableDef detectSchema(String url, String indexName) {
+		SimpleTableDef tableDef = null;
+
+		url += "/schema";
+
+		try {
+			HttpClient httpclient = new DefaultHttpClient();
+			HttpGet request = new HttpGet(url);
+			HttpResponse response = httpclient.execute(request);
+
+			InputStreamReader inputStream = new InputStreamReader(response
+					.getEntity().getContent());
+
+			ObjectReader reader = new ObjectMapper().reader(HashMap.class);
+			Map<String, Object> outerMap = reader.readValue(inputStream);
+
+			Map<String, Object> schemaMap = (Map) outerMap.get("schema");
+
+			if (outerMap != null) {
+				List<Map<String, String>> fieldMapList = (List) schemaMap
+						.get("fields");
+				List<String> columnNames = new ArrayList<String>();
+				List<ColumnType> columnTypes = new ArrayList<ColumnType>();
+
+				for (int i = 0; i < fieldMapList.size(); i++) {
+					Map<String, String> fieldMap = fieldMapList.get(i);
+
+					columnNames.add(fieldMap.get("name"));
+					columnTypes.add(ColumnType.VARCHAR);
```

Comment:
I would suggest using ColumnType.STRING since it's more agnostic (symbolically)... Question: Does Solr not support non-string types?